src/mono/wasm/runtime/logging.ts

```diff
@@ -0,0 +1,207 @@
+                return newRaw;
+        }
+
+        return origMessage;
+    } catch (error) {
+        console.debug(`MONO_WASM: failed to symbolicate: ${error}`);
+        return message;
+    }
+}
+
+export function mono_wasm_stringify_as_error_with_stack(err: Error | string): string {
+    let errObj: any = err;
+    if (!(err instanceof Error))
+        errObj = new Error(err);
+
+    // Error
+    return mono_wasm_symbolicate_string(errObj.stack);
+}
+
+export function mono_wasm_trace_logger(log_domain_ptr: CharPtr, log_level_ptr: CharPtr, message_ptr: CharPtr, fatal: number, user_data: VoidPtr): void {
+    const origMessage = Module.UTF8ToString(message_ptr);
+    const isFatal = !!fatal;
+    const domain = Module.UTF8ToString(log_domain_ptr);
+    const dataPtr = user_data;
+    const log_level = Module.UTF8ToString(log_level_ptr);
+
+    const message = `[MONO] ${origMessage}`;
+
+    if (INTERNAL["logging"] && typeof INTERNAL.logging["trace"] === "function") {
+        INTERNAL.logging.trace(domain, log_level, message, isFatal, dataPtr);
+        return;
+    }
+
+    switch (log_level) {
+        case "critical":
+        case "error":
+            console.error(mono_wasm_stringify_as_error_with_stack(message));
+            break;
+        case "warning":
+            console.warn(message);
+            break;
+        case "message":
+            console.log(message);
+            break;
+        case "info":
+            console.info(message);
+            break;
+        case "debug":
+            console.debug(message);
+            break;
+        default:
+            console.log(message);
+            break;
+    }
+}
+
+export function setup_proxy_console(id: string, console: any, origin: string): void {
+    function proxyConsoleMethod(prefix: string, func: any, asJson: boolean) {
+        return function (...args: any[]) {
+            try {
+                let payload = args[0];
+                if (payload === undefined) payload = "undefined";
+                else if (payload === null) payload = "null";
+                else if (typeof payload === "function") payload = payload.toString();
+                else if (typeof payload !== "string") {
+                    try {
+                        payload = JSON.stringify(payload);
+                    } catch (e) {
+                        payload = payload.toString();
+                    }
+                }
+
+                if (typeof payload === "string")
+                    payload = `[${id}] ${payload}`;
+
+                if (asJson) {
+                    func(JSON.stringify({
+                        method: prefix,
+                        payload: payload,
+                        arguments: args
+                    }));
+                } else {
+                    func([prefix + payload, ...args.slice(1)]);
+                }
+            } catch (err) {
+                originalConsole.error(`proxyConsole failed: ${err}`);
```

Comment:
This is reference `originalConsole` which doesn't seem to be declared yet.

Reply:
We are in the scope of the `proxyConsoleMethod` and the inner function would be called after the `originalConsole` already has a value. I could re-order the code to make it more readable.